File: app/src/main/java/com/xeniac/warrantyrostermanager/landingactivity/forgotpasswordfragment/ForgotPwFragment.java
package com.xeniac.warrantyrostermanager.landingactivity.forgotpasswordfragment;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.Log;
import android.util.Patterns;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;

import com.apollographql.apollo.ApolloCall;
import com.apollographql.apollo.ApolloClient;
import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.exception.ApolloException;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.xeniac.warrantyrostermanager.Constants;
import com.xeniac.warrantyrostermanager.NetworkHelper;
import com.xeniac.warrantyrostermanager.R;
import com.xeniac.warrantyrostermanager.databinding.FragmentForgotPwBinding;

import java.util.Arrays;

public class ForgotPwFragment extends Fragment {

    private FragmentForgotPwBinding forgotPwBinding;
    private View view;
    private Activity activity;
    private Context context;
    private NavController navController;

    public ForgotPwFragment() {
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        forgotPwBinding = FragmentForgotPwBinding.inflate(inflater, container, false);
        view = forgotPwBinding.getRoot();
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        forgotPwBinding = null;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        activity = getActivity();
        context = getContext();
        navController = Navigation.findNavController(view);

        textInputsBackgroundColor();
        textInputsStrokeColor();
        returnOnClick();
        sendOnClick();
        sendActionDone();
    }

    private void textInputsBackgroundColor() {
        forgotPwBinding.tiForgotPwEditEmail.setOnFocusChangeListener((view, focused) -> {
            if (focused) {
                forgotPwBinding.tiForgotPwLayoutEmail.setBoxBackgroundColorResource(R.color.background);
            } else {
                forgotPwBinding.tiForgotPwLayoutEmail.setBoxBackgroundColorResource(R.color.grayLight);
            }
        });
    }

    private void textInputsStrokeColor() {
        forgotPwBinding.tiForgotPwEditEmail.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence inputEmail, int start, int before, int count) {
                forgotPwBinding.tiForgotPwLayoutEmail.setErrorEnabled(false);
                forgotPwBinding.tiForgotPwLayoutEmail.setBoxStrokeColor(context.getResources().getColor(R.color.blue));
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
    }

    private void returnOnClick() {
        forgotPwBinding.btnForgotPwReturn.setOnClickListener(view ->
                activity.onBackPressed());
    }

    private void sendOnClick() {
        forgotPwBinding.btnForgotPwSend.setOnClickListener(view ->
                sendResetPasswordEmail());
    }

    private void sendActionDone() {
        forgotPwBinding.tiForgotPwEditEmail.setOnEditorActionListener((textView, actionId, keyEvent) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                sendResetPasswordEmail();
            }
            return false;
        });
    }

    private void sendResetPasswordEmail() {
        InputMethodManager inputMethodManager = (InputMethodManager)
                context.getSystemService(Context.INPUT_METHOD_SERVICE);
        inputMethodManager.hideSoftInputFromWindow(view.getApplicationWindowToken(), 0);

        if (NetworkHelper.hasNetworkAccess(context)) {
            getResetPasswordInput();
        } else {
            hideLoadingAnimation();
            Snackbar.make(view, context.getResources().getString(R.string.network_error_connection),
                    BaseTransientBottomBar.LENGTH_INDEFINITE)
                    .setAction(context.getResources().getString(R.string.network_error_retry), v -> sendResetPasswordEmail()).show();
        }
    }

    private void getResetPasswordInput() {
        String email;

        if (TextUtils.isEmpty(forgotPwBinding.tiForgotPwEditEmail.getText())) {
            forgotPwBinding.tiForgotPwLayoutEmail.requestFocus();
            forgotPwBinding.tiForgotPwLayoutEmail.setBoxStrokeColor(context.getResources().getColor(R.color.red));
        } else {
            email = forgotPwBinding.tiForgotPwEditEmail.getText().toString().trim().toLowerCase();

            if (!isEmailValid(email)) {
                forgotPwBinding.tiForgotPwLayoutEmail.requestFocus();
                forgotPwBinding.tiForgotPwLayoutEmail.setError(context.getResources().getString(R.string.forgot_pw_error_email));
            } else {
                sendResetPasswordEmailMutation(email);
            }
        }
    }

    private void sendResetPasswordEmailMutation(String email) {
        showLoadingAnimation();

        ApolloClient apolloClient = ApolloClient.builder()
                .serverUrl(Constants.URL_GRAPHQL)
                .build();

        apolloClient.mutate(new SendResetPasswordEmailMutation(email))
                .enqueue(new ApolloCall.Callback<SendResetPasswordEmailMutation.Data>() {
                    @Override
                    public void onResponse(@NonNull Response<SendResetPasswordEmailMutation.Data> response) {
                        if (forgotPwBinding != null) {
                            activity.runOnUiThread(() -> {
                                hideLoadingAnimation();

                                if (!response.hasErrors()) {
                                    Log.i("sendResetPasswordEmail", "onResponse: " + response);
                                    ForgotPwFragmentDirections.ActionForgotPasswordFragmentToForgotPwSentFragment action =
                                            ForgotPwFragmentDirections.actionForgotPasswordFragmentToForgotPwSentFragment(email);
                                    navController.navigate(action);
                                } else {
                                    Log.e("sendResetPasswordEmail", "onResponse Errors: " + response.getErrors());
                                    if (Arrays.toString(response.getErrors().get(0).getCustomAttributes().values().toArray()).contains("code=500")) {
                                        Snackbar.make(view, context.getResources().getString(R.string.forgot_pw_error_not_found),
                                                BaseTransientBottomBar.LENGTH_LONG)
                                                .show();
                                    } else {
                                        Snackbar.make(view, context.getResources().getString(R.string.network_error_response),
                                                BaseTransientBottomBar.LENGTH_INDEFINITE)
                                                .setAction(context.getResources().getString(R.string.network_error_retry),
                                                        v -> sendResetPasswordEmail()).show();
                                    }
                                }
                            });
                        }
                    }

                    @Override
                    public void onFailure(@NonNull ApolloException e) {
                        Log.e("sendResetPasswordEmail", "onFailure: " + e.getMessage());
                        if (forgotPwBinding != null) {
                            activity.runOnUiThread(() -> {
                                hideLoadingAnimation();
                                Snackbar.make(view, context.getResources().getString(R.string.network_error_failure),
                                        BaseTransientBottomBar.LENGTH_LONG).show();
                            });
                        }
                    }
                });
    }

    private void showLoadingAnimation() {
        forgotPwBinding.tiForgotPwEditEmail.setEnabled(false);
        forgotPwBinding.btnForgotPwSend.setClickable(false);
        forgotPwBinding.btnForgotPwSend.setText(null);
        forgotPwBinding.cpiForgotPw.setVisibility(View.VISIBLE);
    }

    private void hideLoadingAnimation() {
        forgotPwBinding.cpiForgotPw.setVisibility(View.GONE);
        forgotPwBinding.tiForgotPwEditEmail.setEnabled(true);
        forgotPwBinding.btnForgotPwSend.setClickable(true);
        forgotPwBinding.btnForgotPwSend.setText(context.getResources().getString(R.string.login_btn_login));
    }

    private boolean isEmailValid(String email) {
        return Patterns.EMAIL_ADDRESS.matcher(email).matches();
    }
}

File: app/src/main/java/com/xeniac/warrantyrostermanager/landingactivity/forgotpasswordfragment/ForgotPwFragment.kt
package com.xeniac.warrantyrostermanager.landingactivity.forgotpasswordfragment

import android.content.Context
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.util.Patterns
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.navigation.NavController
import androidx.navigation.Navigation
import com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_INDEFINITE
import com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_LONG
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.xeniac.warrantyrostermanager.NetworkHelper
import com.xeniac.warrantyrostermanager.R
import com.xeniac.warrantyrostermanager.databinding.FragmentForgotPwBinding
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext

class ForgotPwFragment : Fragment(R.layout.fragment_forgot_pw) {

    private var _binding: FragmentForgotPwBinding? = null
    private val binding get() = _binding!!
    private lateinit var navController: NavController

    private lateinit var firebaseAuth: FirebaseAuth

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        _binding = FragmentForgotPwBinding.bind(view)
        navController = Navigation.findNavController(view)
        firebaseAuth = FirebaseAuth.getInstance()

        textInputsBackgroundColor()
        textInputsStrokeColor()
        returnOnClick()
        sendOnClick()
        sendActionDone()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun textInputsBackgroundColor() {
        binding.tiForgotPwEditEmail.setOnFocusChangeListener { _, isFocused ->
            if (isFocused) {
                binding.tiForgotPwLayoutEmail.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.background)
            } else {
                binding.tiForgotPwLayoutEmail.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.grayLight)
            }
        }
    }

    private fun textInputsStrokeColor() {
        binding.tiForgotPwEditEmail.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(
                inputEmail: CharSequence?,
                start: Int,
                before: Int,
                count: Int
            ) {
                binding.tiForgotPwLayoutEmail.isErrorEnabled = false
                binding.tiForgotPwLayoutEmail.boxStrokeColor =
                    ContextCompat.getColor(requireContext(), R.color.blue)
            }

            override fun afterTextChanged(s: Editable?) {
            }
        })
    }

    private fun returnOnClick() {
        binding.btnForgotPwReturn.setOnClickListener {
            requireActivity().onBackPressed()
        }
    }

    private fun sendOnClick() {
        binding.btnForgotPwSend.setOnClickListener {
            sendResetPasswordEmail()
        }
    }

    private fun sendActionDone() {
        binding.tiForgotPwEditEmail.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                sendResetPasswordEmail()
            }
            false
        }
    }

    private fun sendResetPasswordEmail() {
        val inputMethodManager = requireContext()
            .getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        inputMethodManager.hideSoftInputFromWindow(binding.root.applicationWindowToken, 0)

        if (NetworkHelper.hasNetworkAccess(requireContext())) {
            getResetPasswordInput()
        } else {
            hideLoadingAnimation()
            Snackbar.make(
                binding.root,
                requireContext().getString(R.string.network_error_connection),
                LENGTH_INDEFINITE
            ).apply {
                setAction(requireContext().getString(R.string.network_error_retry)) { sendResetPasswordEmail() }
                show()
            }
        }
    }

    private fun getResetPasswordInput() {
        val email = binding.tiForgotPwEditEmail.text.toString().trim().lowercase()

        if (email.isBlank()) {
            binding.tiForgotPwLayoutEmail.requestFocus()
            binding.tiForgotPwLayoutEmail.boxStrokeColor =
                ContextCompat.getColor(requireContext(), R.color.red)
        } else {
            if (!isEmailValid(email)) {
                binding.tiForgotPwLayoutEmail.requestFocus()
                binding.tiForgotPwLayoutEmail.error =
                    requireContext().getString(R.string.forgot_pw_error_email)
            } else {
                sendResetPasswordEmailAuth(email)
            }
        }
    }

    private fun sendResetPasswordEmailAuth(email: String) {
        showLoadingAnimation()

        CoroutineScope(Dispatchers.IO).launch {
            try {
                firebaseAuth.sendPasswordResetEmail(email).await()
                Log.i(
                    "sendResetPasswordEmail",
                    "Reset password email successfully sent to ${email}."
                )
                withContext(Dispatchers.Main) {
                    hideLoadingAnimation()
                    val action = ForgotPwFragmentDirections
                        .actionForgotPasswordFragmentToForgotPwSentFragment(email)
                    navController.navigate(action)
                }
            } catch (e: Exception) {
                Log.e("sendResetPasswordEmail", "Exception: ${e.message}")
                withContext(Dispatchers.Main) {
                    hideLoadingAnimation()
                    if (e.toString().contains(
                            "There is no user record corresponding to this identifier"
                        )
                    ) {
                        Snackbar.make(
                            binding.root,
                            requireContext().getString(R.string.forgot_pw_error_not_found),
                            LENGTH_LONG
                        ).show()
                    } else {
                        Snackbar.make(
                            binding.root,
                            requireContext().getString(R.string.network_error_failure),
                            LENGTH_LONG
                        ).show()
                    }
                }
            }
        }
    }

    private fun showLoadingAnimation() {
        binding.tiForgotPwEditEmail.isEnabled = false
        binding.btnForgotPwSend.isClickable = false
        binding.btnForgotPwSend.text = null
        binding.cpiForgotPw.visibility = VISIBLE
    }

    private fun hideLoadingAnimation() {
        binding.tiForgotPwEditEmail.isEnabled = true
        binding.btnForgotPwSend.isClickable = true
        binding.btnForgotPwSend.text =
            requireContext().getString(R.string.login_btn_login)
        binding.cpiForgotPw.visibility = GONE
    }

    private fun isEmailValid(email: String): Boolean {
        return Patterns.EMAIL_ADDRESS.matcher(email).matches()
    }
}